src/object/obj_class.c

```diff
@@ -328,3 +356,127 @@ daos_oclass_grp_nr(struct daos_oclass_attr *oc_attr, struct daos_obj_md *md)
 	/* NB: @md is unsupported for now */
 	return oc_attr->ca_grp_nr;
 }
+
+/** a structure to map EC object class to gf tables */
+struct daos_oc_gftbls {
+	/** object class id */
+	daos_oclass_id_t	 e_oc_id;
+	/**
+	 * pointer to array of input tables generated from coding coefficients.
+	 */
+	unsigned char		*e_gftbls;
+};
+
+static struct daos_oc_gftbls	*daos_oc_gftbls_map;
+static int			 daos_oc_gftbls_nr;
+
+void
+daos_oc_gftbls_map_fini()
+{
+	struct daos_oc_gftbls	*oc_gftbls;
+	struct daos_obj_class	*oc;
+	int			 ocnr = 0;
+	int			 i;
+
+	if (daos_oc_gftbls_map == NULL)
+		return;
+
+	for (oc = &daos_obj_classes[0]; oc->oc_id != DAOS_OC_UNKNOWN; oc++) {
+		if (oc->oc_attr.ca_resil == DAOS_RES_EC)
+			ocnr++;
+	}
+	D_ASSERT(daos_oc_gftbls_nr == ocnr);
+
+	for (i = 0; i < ocnr; i++) {
+		oc_gftbls = &daos_oc_gftbls_map[i++];
+		if (oc_gftbls->e_gftbls != NULL)
+			D_FREE(oc_gftbls->e_gftbls);
+	}
+
+	D_FREE(daos_oc_gftbls_map);
+	daos_oc_gftbls_map = NULL;
+	daos_oc_gftbls_nr = 0;
+}
+
+int
+daos_oc_gftbls_map_init()
+{
+	struct daos_oc_gftbls	*oc_gftbls;
+	struct daos_obj_class	*oc;
+	unsigned char		*encode_matrix = NULL;
+	int			 ocnr = 0;
+	int			 i;
+	int			 k, p, m;
+	int			 rc;
+
+	if (daos_oc_gftbls_map != NULL)
+		return 0;
+
+	for (oc = &daos_obj_classes[0]; oc->oc_id != DAOS_OC_UNKNOWN; oc++) {
+		if (oc->oc_attr.ca_resil == DAOS_RES_EC)
+			ocnr++;
+	}
+
+	if (ocnr == 0)
+		return 0;
+
+	D_ALLOC_ARRAY(daos_oc_gftbls_map, ocnr);
+	if (daos_oc_gftbls_map == NULL)
+		D_GOTO(failed, rc = -DER_NOMEM);
+
+	i = 0;
+	for (oc = &daos_obj_classes[0]; oc->oc_id != DAOS_OC_UNKNOWN; oc++) {
+		if (oc->oc_attr.ca_resil != DAOS_RES_EC)
+			continue;
+
+		oc_gftbls = &daos_oc_gftbls_map[i++];
+		oc_gftbls->e_oc_id = oc->oc_id;
+		k = oc->oc_attr.u.ec.e_k;
+		p = oc->oc_attr.u.ec.e_p;
+		m = k + p;
+		D_ALLOC(oc_gftbls->e_gftbls, k * p * 32);
+		if (oc_gftbls->e_gftbls == NULL)
+			D_GOTO(failed, rc = -DER_NOMEM);
+		D_ALLOC(encode_matrix, m * k);
+		if (encode_matrix == NULL)
+			D_GOTO(failed, rc = -DER_NOMEM);
+		/* A Cauchy matrix is always invertible, the recovery rule is
+		 * simpler than gf_gen_rs_matrix (vandermonde matrix).
+		 */
+		gf_gen_cauchy1_matrix(encode_matrix, m, k);
+		/* Initialize gf tables from encode matrix */
+		ec_init_tables(k, p, &encode_matrix[k * k],
+			       oc_gftbls->e_gftbls);
```

Comment:
the max allowed K/P value for GF(2^8) is 255.
init as needed will add some overhead for each OP, since it only related with K/P parameter and cauchy/RS matrix algorithm, just use another table to init/cache it one time seems better to me.

Reply:
yep, I'm fine with initializing at the begin, keep it simple if cost is not high